src/refresh/RefreshTimer.ts

```diff
@@ -0,0 +1,106 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+/**
+ * The backoff time is between the minimum and maximum backoff time, based on the number of attempts.
+ * An exponential backoff strategy is used, with a jitter factor to prevent clients from retrying at the same time.
+ *
+ * The backoff time is calculated as follows:
+ *  - `basic backoff time` = `MinimumBackoffInMs` * 2 ^ `attempts`, and it is no larger than the `MaximumBackoffInMs`.
+ *  - based on jitter ratio, the jittered time is between [-1, 1) * `JitterRatio` * basic backoff time.
+ *  - the final backoff time is the basic backoff time plus the jittered time.
+ *
+ * Note: the backoff time usually is no larger than the refresh interval, which is specified by the user.
+ *  - If the interval is less than the minimum backoff, the interval is used.
+ *  - If the interval is between the minimum and maximum backoff, the interval is used as the maximum backoff.
+ *  - Because of the jitter, the maximum backoff time is actually `MaximumBackoffInMs` * (1 + `JitterRatio`).
+ */
+
+const MinimumBackoffInMs = 30 * 1000; // 30s
+const MaximumBackoffInMs = 10 * 60 * 1000; // 10min
+const MaxSafeExponential = 53; // Used to avoid overflow, as Number.MAX_SAFE_INTEGER = 2^53 - 1.
+const JitterRatio = 0.25;
+
+export class RefreshTimer {
+    private _minBackoff: number;
+    private _maxBackoff: number;
+    private _attempts: number;
+    private _backoffEnd: number; // Timestamp
+    constructor(
+        private _interval: number
+    ) {
+        if (this._interval <= 0) {

```

Comment:
The `Math.min()` here can be removed. In `_CalculateBackoffTime()`, we have `if (this._interval <= this._minBackoff)`

Reply:
Good catch. It's duplicate.

Since the logic is one-time initialization given fixed interval and min/max backoff time, I think it's better to move this part into constructor instead of in `_calculateBackoffTime`. It avoids repeated calculations.

Reply:
Let me first remove the duplicate code from constructor. let's see how it looks like.